### PR TITLE
Price discovery improvements

### PR DESCRIPTION
**Pull request by dorin-iancu**
- owner can now select a min-price for the launched token. The check is only activated after the initial launched tokens are deposited
- add an option for owner to add extra rewards. Users will receive their share based on the ratio of lp tokens the are entitled to.
- fix extra lp tokens calculation to also use the ratio of lp tokens from total

### CODE REVIEW

File: price-discovery/src/lib.rs [hunk 1]
```diff
@@ -243,4 +290,27 @@ pub trait PriceDiscovery:
...
+        let launched_token_balance = self.blockchain().get_sc_balance(&launched_token_id, 0);
+        let accepted_token_balance = self.blockchain().get_sc_balance(&accepted_token_id, 0);
+
+        if accepted_token_balance == 0 {
+            return;
+        }
+
+        require!(launched_token_balance > 0, "No launched tokens available");
+        require!(
+            accepted_token_balance * MIN_PRICE_PRECISION / launched_token_balance >= min_price,
```
Comment (claudiulataretu) [inline price-discovery/src/lib.rs:308]: maybe extract this price compute in a variable

**Review by claudiu725 — DISMISSED** [review]

File: price-discovery/src/lib.rs (hunk 1 above)
Comment (claudiulataretu) [reply price-discovery/src/lib.rs:308]: I think `MIN_PRICE_PRECISION` should be set based on `launched_token` decimals numbers

Comment (sasurobert) [reply price-discovery/src/lib.rs:308]: This can be changed on deploy - I do not see it necessary to be in config or kept in storage

**Review by sasurobert — APPROVED** [review]
Looks good!

File: price-discovery/src/lib.rs [hunk 2]
```diff
@@ -185,49 +225,56 @@ pub trait PriceDiscovery:
...
-            self.get_percentage_of_total_supply(redeem_token_nonce, redeem_token_amount);
-
-        let extra_lp_tokens = self.extra_lp_tokens().get();
-        let bonus = percentage_of_redeeem_token_supply * extra_lp_tokens / MAX_PERCENTAGE / 2u32;
-        self.extra_lp_tokens()
-            .update(|extra_amt| *extra_amt -= &bonus);
-
-        match redeem_token_nonce {
+    ) -> RewardsPair<Self::Api> {
+        let total_lp_tokens = self.total_claimable_lp_tokens().get();
```
Comment (sasurobert) [inline price-discovery/src/lib.rs:256]: total_lp_received would be better here.

File: price-discovery/src/lib.rs [hunk 3]
```diff
@@ -185,49 +225,56 @@ pub trait PriceDiscovery:
...
-                let base_lp_amount =
-                    redeem_token_amount * &total_lp_tokens / accepted_token_final_amount / 2u32;
-
-                base_lp_amount + bonus
+                redeem_token_amount * &total_lp_tokens / accepted_token_final_amount / 2u32
             }
             _ => BigUint::zero(),
+        };
+
+        let total_extra_lp_tokens = self.extra_lp_tokens().get();
```
Comment (sasurobert) [inline price-discovery/src/lib.rs:269]: there is no extra lp. as everything was put into swap SC

**Review by sasurobert — CHANGES_REQUESTED** [review]
for above comments

**Review by claudiu725 — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**